### ISSUE TITLE
CA1859: false positive if the value is changed by await

### ISSUE THREAD

**Issue by pcf0** [dotnet/roslyn-analyzers#6904]
### Analyzer

**Diagnostic ID**: [CA1859](https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1859): `Use concrete types when possible for improved performance`

### Analyzer source

**SDK**: [Built-in CA analyzers in .NET 8 SDK or later](https://learn.microsoft.com/dotnet/fundamentals/productivity/code-analysis)

**Version**: [SDK 8.0.100](https://dotnet.microsoft.com/download/dotnet/8.0)

_OR_

**NuGet Package**: [Microsoft.CodeAnalysis.NetAnalyzers](https://www.nuget.org/packages/Microsoft.CodeAnalysis.NetAnalyzers)

**Version**: 8.0.0 (Latest 2023-11-17)

### Describe the bug

A property or field with an interface type is initialized with a concrete implementation. In an `async` method another implementation is assigned using `await`. CA1859 is generated.

### Steps To Reproduce
```csharp
public class Class1
{
    private I Prop { get; set; } = new Impl1(); //<-- CA1859 

    public async Task Init()
    {
        Prop = await Task.FromResult<I>(new Impl2());
        Prop.M();
    }
}

internal interface I
{
    void M();
}

internal class Impl1 : I
{
    public void M() { }
}

internal class Impl2 : I
{
    public void M() { }
}
```

### Expected behavior

No errors reported

### Actual behavior

Error reported:
`Class1.cs(3,15): error CA1859: Change type of property 'Prop' from 'I' to 'Impl1' for improved performance (https://learn.microsoft.com/dotnet/fundamentals/code-analysis/quality-rules/ca1859)`

**Comment by JustArchi** [comment]
Another repro for this on rc1:

```csharp
	public static async Task<IList<T>> InParallel2<T>(IEnumerable<Task<T>> tasks) {
		ArgumentNullException.ThrowIfNull(tasks);

		IList<T> results;

		switch (DateTime.Now.Year) {
			case 2023:
				results = new List<T>();

				foreach (Task<T> task in tasks) {
					results.Add(await task.ConfigureAwait(false));
				}

				break;
			default:
				results = await Task.WhenAll(tasks).ConfigureAwait(false);

				break;
		}

		return results;
	}
```
